apollo-federation/src/connectors/json_selection/methods/public/filter.rs

```diff
@@ -0,0 +1,293 @@
+/// $->filter(@->eq(3))      result is [3]
+/// $->filter(@->gt(3))      result is [4, 5]
+///
+/// We are taking each value passed into filter via @ and running the condition function against that value.
+/// Only values where the condition returns true will be included in the result array.
+///
+/// Example with objects:
+/// users->filter(@.active->eq(true))    returns only users where active is true
+fn filter_method(
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    data: &JSON,
+    vars: &VarsWithPathsMap,
+    input_path: &InputPath<JSON>,
+) -> (Option<JSON>, Vec<ApplyToError>) {
+    let Some(args) = method_args else {
+        return (
+            None,
+            vec![ApplyToError::new(
+                format!("Method ->{} requires one argument", method_name.as_ref()),
+                input_path.to_vec(),
+                method_name.range(),
+            )],
+        );
+    };
+    let Some(first_arg) = args.args.first() else {
+        return (
+            None,
+            vec![ApplyToError::new(
+                format!("Method ->{} requires one argument", method_name.as_ref()),
+                input_path.to_vec(),
+                method_name.range(),
+            )],
+        );
+    };
+
+    if let JSON::Array(array) = data {
+        let mut output = Vec::new();
+        let mut errors = Vec::new();
+        let mut has_non_boolean_error = false;
+
+        for (i, element) in array.iter().enumerate() {
+            let input_path = input_path.append(JSON::Number(i.into()));
+            let (applied_opt, arg_errors) = first_arg.apply_to_path(element, vars, &input_path);
+            errors.extend(arg_errors);
+
+            match applied_opt {
+                Some(JSON::Bool(true)) => {
+                    output.push(element.clone());
+                }
+                Some(JSON::Bool(false)) | None => {
+                    // Condition is false or errored, exclude the element
+                }
+                Some(_) => {
+                    // Condition returned a non-boolean value, this is an error
+                    has_non_boolean_error = true;
+                    errors.push(ApplyToError::new(
+                        "Filter condition must return a boolean value".to_string(),
+                        input_path.to_vec(),
+                        method_name.range(),
+                    ));
+                }
+            }
+        }
+
+        if has_non_boolean_error {
+            (None, errors)
+        } else {
+            (Some(JSON::Array(output)), errors)
+        }
+    } else {
+        // For non-array inputs, treat as single-element array
+        // Apply filter condition and return either [value] or []
+        let (condition_result, mut condition_errors) =
+            first_arg.apply_to_path(data, vars, input_path);
+
+        match condition_result {
+            Some(JSON::Bool(true)) => (Some(JSON::Array(vec![data.clone()])), condition_errors),
```

Comment:
if this is not an array, why treat it as an array?

Languages like Rust, OCaml, Clojure, Haskell have pretty interesting ways of dealing with non-array filtering. Maybe we should consider a broader approach to this case and return the value itself or none?

https://doc.rust-lang.org/std/option/enum.Option.html#method.filter